mtp/mtp_fs.cpp

```diff
@@ -4,336 +4,320 @@
+        // TODO: list dir is not a good choose, as it allocates
+        // memory for all files from directory
+        // auto list = vfs.listdir(root);
+        // it would be better to have iterator mapped to
+        // filesystem.
+
+        if (root != 0 && root != 0xFFFFFFFF) {
+            return 0;
+        }
 
-static uint64_t get_free_space(void *arg)
-{
-    // TODO: see get_disk_properties
-    uint64_t size = 0;
-    struct statvfs stvfs {};
-    statvfs( purefs::dir::getUserDiskPath().c_str(), &stvfs);
-
-    size = (uint64_t(stvfs.f_bsize) * stvfs.f_bavail) - getSpareSpace();
-    LOG("Free space: %u MB", unsigned(size / 1024 / 1024));
-    return size;
-}
+        fs->find_data = opendir(fs->root);
+        if (fs->find_data == nullptr) {
+            log_error("Opendir failed");
+            return 0;
+        }
+        *count = count_files(fs->find_data);
+        log_debug("Found: %u files", static_cast<unsigned>(*count));
+        if (*count == 0) {
+            return 0; // empty directory
+        }
 
-static uint32_t fs_find_first(void *arg, uint32_t root, uint32_t *count)
-{
-    struct mtp_fs *fs = (struct mtp_fs*)arg;
-    uint32_t handle;
-    // TODO: list dir is not a good choose, as it allocates
-    // memory for all files from directory
-    // auto list = vfs.listdir(ROOT);
-    // it would be better to have iterator mapped to
-    // filesystem.
-
-    if (root != 0 && root != 0xFFFFFFFF)
-        return 0;
+        rewinddir(fs->find_data);
+        struct dirent *de;
+        while (((de = readdir(fs->find_data)) != nullptr) && is_dot(de->d_name)) {
+            log_debug("Skip: '%s'", de->d_name);
+        }
+        if (de == nullptr) {
+            log_error("No files found");
+            *count = 0;
+            return 0;
+        }
+        handle = mtp_db_add(fs->db, de->d_name);
+        return handle;
+    }
 
-    fs->find_data = opendir(fs->ROOT);
-    if(!fs->find_data)
+    uint32_t fs_find_next(void *arg)
     {
-        LOG("Opendir failed");
+        uint32_t handle;
+        struct mtp_fs *fs = static_cast<struct mtp_fs *>(arg);
+        struct dirent *de;
+        while ((de = readdir(fs->find_data)) != nullptr) {
+            if (is_dot(de->d_name)) {
+                continue;
+            }
+            handle = mtp_db_add(fs->db, de->d_name);
+            return handle;
+        }
+        log_debug("Done, no more files");
         return 0;
     }
-    *count = count_files(fs->find_data);
 
-    // empty directory
-    if (*count == 0)
-        return 0;
-
-    LOG("found: %u files:", (unsigned int) *count);
-    rewinddir(fs->find_data);
-    struct dirent *de;
-    while( (de=readdir(fs->find_data)) &&  is_dot(de->d_name)) {
-        LOG("skip: %s", de->d_name);
-    }
-    if( !de )
+    uint16_t ext_to_format_code(const char *name)
     {
-        LOG("No files found");
-        *count = 0;
-        return 0;
+        if (strstr(".jpg", name) != nullptr) {
```

Comment:
This way of checking the file extension can be unreliable, see https://github.com/mudita/MuditaOS/pull/4121#discussion_r1003080987.

Reply:
Fixed

Reply:
I think that .jp**e**g is also a valid extension for this format.

Reply:
You're right, added